### PR TITLE
Add touch support in our carousel

### PR DESCRIPTION
**Pull request by Johann-S**
- [x] Add unit test
- [x] Documentation

Fixes: #17118

### CODE REVIEW

**Comment by XhmikosR** [comment]
Its quite a big addition... Can't we do the basic stuff ourselves?

File: _includes/scripts.html
Comment (XhmikosR) [inline _includes/scripts.html]: You need to add it in package.json too for the docs package.

**Comment by Johann-S** [comment]
I don't think it's a big change here, we add an optionnal lib, if folks need a touch support that's all

**Comment by XhmikosR** [comment]
Sure, but can't we have the basic functionality without a 20 KB file? I recall you had a branch in the past.

**Comment by Johann-S** [comment]
This won't be added in our dist files, so we won't add 20 kb

I'm not in favor of adding logics about touch detection and swipe detection when they are libs which can do that better than me

**Comment by XhmikosR** [comment]
I agree on that, but then one would say why don't you support jQuery mobile or whatever other lib there's out there.

Anyway, I'm not against this, I just want us to think it through.

**Comment by Johann-S** [comment]
I think HammerJS is a good choice, Angular chose HammerJS too

File: _includes/scripts.html
Comment (Johann-S) [reply _includes/scripts.html]: It's strange but Popper isn't in our `doc-minify` task 🤔

Comment (XhmikosR) [reply _includes/scripts.html]: True, indeed. Just leave it as is and I'll give it a go in the next days and let you know. :)

File: js/src/carousel.js
Comment (jackmu95) [inline js/src/carousel.js]: Isn't here a check quired if hammer support is enabled or the hammer reference exists to prevent a reference error? 🤔

Comment (Johann-S) [reply js/src/carousel.js]: Yep that's true nice catch 👍

File: js/tests/index.html
Comment (Johann-S) [reply js/tests/index.html]: yep sure !

File: js/tests/karma.conf.js
Comment (XhmikosR) [inline js/tests/karma.conf.js]: This shouldn't be needed.

**Comment by XhmikosR** [comment]
@Johann-S can you rebase this?

**Comment by Johann-S** [comment]
Done @XhmikosR 👍

**Comment by Johann-S** [comment]
I cannot validate our test because on SauceLabs we run our test on Linux Firefox 45, Linux Chrome 48 and Mac OS X 10.11 Safari 10 which are obsolete version

**Comment by XhmikosR** [comment]
@Johann-S: so we can't merge this now if we can't test it, right? What can we do about this?

**Comment by XhmikosR** [comment]
@mdo: can this be done in 4.2?

**Comment by Johann-S** [comment]
@XhmikosR everything is done here and works, thanks to BrowserStack 😄 

/CC @bardiharborow for a review

File: js/tests/unit/carousel.js
Comment (patrickhlauke) [inline js/tests/unit/carousel.js]: for testing purposes, this is of course fine. i'll just note here that to detect pointer events touch support you'd want to also check for `navigator.maxTouchPoints > 0` (in production on a live site, not for automated unit tests)

File: site/docs/4.1/components/carousel.md
Comment (patrickhlauke) [inline site/docs/4.1/components/carousel.md]: perhaps reword this a tad:

"The carousel supports swipe gestures (left and right) using <a href="https://hammerjs.github.io/">HammerJS</a>. For this to function correctly you need to include [HammerJS]...."

File: js/tests/unit/carousel.js
Comment (Johann-S) [reply js/tests/unit/carousel.js]: yep, I added that line because on Safari our unit tests with touch events failed 😄

File: site/docs/4.1/components/carousel.md
Comment (Johann-S) [reply site/docs/4.1/components/carousel.md]: I changed that, thanks Patrick !

**Comment by XhmikosR** [comment]
@Johann-S: can you rebase this one please? :)

**Comment by Johann-S** [comment]
Done @XhmikosR 😉

**Comment by XhmikosR** [comment]
All good with this @Johann-S. Only thing I was thinking is if we should have an option to enable/disable touch support.

**Comment by Johann-S** [comment]
If they don't include HammerJS there is no touch support

**Comment by XhmikosR** [comment]
Yes, but the bundle already includes it.

File: js/src/carousel.js
Comment (Johann-S) [inline js/src/carousel.js]: I should use hammer only if touch events supported

**Comment by ben29** [comment]
So when it will be supported?

**Comment by XhmikosR** [comment]
@Johann-S: would you be so kind and rebase this one last time (hopefully)?

**Comment by Johann-S** [comment]
Done @XhmikosR 😉 Maybe someday it'll be merged 🙏

**Review by XhmikosR — APPROVED** [review]

**Comment by XhmikosR** [comment]
OK, I think this is ready to be merged :)

File: site/docs/4.1/components/carousel.md
Comment (XhmikosR) [inline site/docs/4.1/components/carousel.md]: One thing. This is true, but maybe we should mention that it will be disabled gracefully if HammerJs isn't present?

Comment (Johann-S) [reply site/docs/4.1/components/carousel.md]: You're right we should add that

**Review by patrickhlauke — DISMISSED** [review]
the documentation part may need some more clarity, as it's not quite obvious to me what that even means...

File: site/docs/4.1/components/carousel.md
Comment (patrickhlauke) [inline site/docs/4.1/components/carousel.md]: no, not really "only mobile browsers" - there are desktop computers with touch (e.g. surface). also, hammer uses pointer events where available, so technically not "touch events". suggest changing it to what i said originally in my follow-up comment :)

**Review by patrickhlauke — CHANGES_REQUESTED** [review]

**Comment by patrickhlauke** [comment]
do any of the demo carousels on https://deploy-preview-25776--twbs-bootstrap4.netlify.com/docs/4.1/components/carousel/ have touch enabled? if not, would be good to do it and note it

**Comment by Johann-S** [comment]
All the carousels have touch enabled 👍

File: site/docs/4.1/components/carousel.md
Comment (Johann-S) [reply site/docs/4.1/components/carousel.md]: Oh ok I'll change that. We removed HammerJS we use a Vanilla JS solution

Comment (patrickhlauke) [reply site/docs/4.1/components/carousel.md]: oh, missed that you're using vanilla JS (as i saw the hammer simulator being loaded as well). in which case could I'd say it would be good to also support pointer events...but that's probably a more involved addition...

File: js/src/carousel.js
Comment (patrickhlauke) [inline js/src/carousel.js]: if you're not using pointer events, why are you testing for `navigator.maxTouchPoints` (which is part of the pointer events spec) ?

File: site/docs/4.1/components/carousel.md
Comment (patrickhlauke) [reply site/docs/4.1/components/carousel.md]: for reference: supporting pointer events will also make this work on touch-enabled desktop/laptop devices in Edge (which, by default, does not expose touch events by default, but does support touch interactions via pointer events)

**Comment by patrickhlauke** [comment]
swiping left/right in Chrome on desktop/laptop with touchscreen and on android also triggers back/forward navigation. it likely needs a preventDefault() in there somewhere (this is the sort of stuff that i think Hammer would have handled by default already, as well as pointer events support)

**Comment by XhmikosR** [comment]
>the swipe direction seems counterintuitive. if i swipe right, i would want the carousel to go to the previous slide, not the next slide, and vice-versa

Yeah, that was one of the things that didn't feel right to me either. I asked Johann to keep the HammerJS patch just in case we need to revert. So if things tend to be too broken, we can still just go with Hammer.

**Review by patrickhlauke — APPROVED** [review]

**Review by patrickhlauke — CHANGES_REQUESTED** [review]
- revert the swipe direction (so swiping left goes to next slide, not previous, and vice-versa)
- prevent swipes from triggering browser back/forward navigation (e.g. Chrome on Android / on Surface)

**Comment by Johann-S** [comment]
it should be better @patrickhlauke can you confirm it ? Please

**Review by patrickhlauke — APPROVED** [review]

**Comment by patrickhlauke** [comment]
strangely, it still doesn't work in Edge on Surface (while the hammer.js demo itself http://hammerjs.github.io/ does). i'll have a look later tonight to see if i can work out why. other than than, the swipe direction now feels natural, and back/forward in Chrome is suppressed when swiping. :+1:

**Comment by Johann-S** [comment]
I tested (thanks to BrowserStack) on a Lumia 930 with Windows Phone 8.1 and it works fine 🤔

**Comment by patrickhlauke** [comment]
Win Phone 8 behaves differently from a desktop Windows 10 on a touchscreen.

**Comment by patrickhlauke** [comment]
ah, i see you had added hammer, but then removed it again, hence my confusion. to be clear: Edge/Win 10 on a touch-enabled desktop/laptop does not have touch events enabled, but does allow touch interaction via pointer events (filtered to `type = "touch"`, and with addition of relevant `touch-action: pan-x` to the element/container with the event handlers). as you reverted to only touch events again after removing hammer, it's not surprising it's not working in Edge on Surface. i'd say although it's an edge case (hah) at this point, since pointer events are the more modern way of handling things (and finally even Apple have them "in development" for future Safari), it's worth future-proofing this code here as well. i can give it a shot if you don't mind me fuzzing around with your branch?

**Comment by patrickhlauke** [comment]
should ideally use both, with preference for PE where available. i'll have a go either later tonight or tomorrow

**Comment by Johann-S** [comment]
I implemented a solution with PointerEvent (finally I found the time 😄) but I'm not sure if I'm fully compatible 😟 
If you can help me on that @patrickhlauke or improve my code do not hesitate 👍

File: js/src/carousel.js
Comment (patrickhlauke) [inline js/src/carousel.js]: should this not only be added if the configuration for touch is true?

File: js/src/carousel.js
Comment (patrickhlauke) [inline js/src/carousel.js]: beyond adding the event listener, this code should also add `touch-action: pan-x` to the carousel when pointer events are used ... `preventDefault()` doesn't stop the forward/back browser navigation for pointer events, it's done declaratively with `touch-action` in the PE model. either doing it dynamically (adding a `style`? seems a bit dirty) or adding it to the definition of carousel CSS styles (though then it would "swallow" horizontal swipes in a PE browser even if the author DIDN'T want touch to be handled on a carousel)

File: js/src/carousel.js
Comment (patrickhlauke) [inline js/src/carousel.js]: ideally also check first that the pointer event is of `pointerType=="touch"` - and possibly even `pointerType="pen"` if we want to allow pen/stylus swiping - as otherwise the swipe will also happen for mouse users. see https://w3c.github.io/pointerevents/#dom-pointerevent-pointertype

File: js/tests/unit/carousel.js
Comment (patrickhlauke) [inline js/tests/unit/carousel.js]: does this only simulate/test touch events? or does it also test if the code works correctly for pointer events?

File: js/src/carousel.js
Comment (Johann-S) [reply js/src/carousel.js]: I added a class when we use PointerEvents but I don't understand why it doesn't work on IE 10/11, do you any ideas why @patrickhlauke ? 🤔

**Comment by patrickhlauke** [comment]
urgh, i'm an idiot . the touch action that's needed isn't `touch-action: pan-x` ... i seem to always forget that `touch-action` really means "which actions should the *browser* handle", and horizontal panning is exactly what we *don't* want the browser to handle...

long story short, changing to `touch-action: pan-y pinch-zoom` makes this work in Edge/Surface and IE11/Surface as well

File: js/src/carousel.js
Comment (patrickhlauke) [reply js/src/carousel.js]: i'm an idiot...and suggested the wrong `touch-action` value...i should go and have a word with the co-editor and chair of the PE spec, it seems...

**Comment by XhmikosR** [comment]
So in Firefox 62 dev tools emulation now it throws this:

```
TypeError: originEvent.touches is undefined carousel.js:287:8
start
carousel.js:287:8
_addTouchEventListeners/<
carousel.js:329:56
```

**Comment by Johann-S** [comment]
It seems not @patrickhlauke our unit tests do not pass on IE 10/11 😟

**Comment by Johann-S** [comment]
@patrickhlauke I cleaned a bit the git history and squashed some of your commits

**Comment by Johann-S** [comment]
@patrickhlauke everything is fine now, can you confirm it ? Thanks 👍

File: js/src/carousel.js
Comment (patrickhlauke) [inline js/src/carousel.js]: the original check for `navigator.maxTouchPoints > 0` was, i'd say, more appropriate. as pointer events (in modern browsers) are supported/present even if there's just a mouse, so you end up with unnecessary code being executed...unless you're doing this on purpose to cover the scenario where an (external) touch-capable display is attached at a later point after the page has already loaded, which is a bit niche...

Comment (patrickhlauke) [reply js/src/carousel.js]: i guess if you did revert back to the `navigator.maxTouchPoints` check, you'd then need to also fake it in the unit tests so that that code gets run

Comment (Johann-S) [reply js/src/carousel.js]: The unit tests wasn't in success because when you swipe on a Imagine with IE it start a drag event 😬 I'll to check maxTouchPoints and see if everything is fine

Comment (Johann-S) [reply js/src/carousel.js]: Ok I restored the use of `navigator.maxTouchPoints`, it should be fine now 👍 I'll make some tests on mobile devices

**Comment by Johann-S** [comment]
It works fine for me 👍 

/CC @XhmikosR if you want to do some tests too

**Comment by patrickhlauke** [comment]
https://deploy-preview-25776--twbs-bootstrap4.netlify.com/docs/4.1/components/carousel/ seems to be working fine now (tested there as i've not got my build/test environment all set up again at the moment) ... but man, the aggressive caching/service worker really gets in the way of quickly checking changes out sometimes...

possibly more related to netlify here: on IE11 and Edge, I found that the carousels are unresponsive for quite a long time after page load (unless you kick them into action if there are the "<" / ">" controls). it seems the browser is waiting for some file to load from the network, and only once it gives up does it run the onload stuff.

**Comment by XhmikosR** [comment]
There is actually no service worker on netlify; it doesn't run the production scripts on purpose.

**Comment by patrickhlauke** [comment]
> There is actually no service worker on netlify; it doesn't run the production scripts on purpose.

i found i had to force refresh (and in the case of Firefox, go in and delete session storage etc) while testing. not quite sure why then...aggressive caching? i see pwa.js is loaded as well if that means anything...i'm actually not sure if this is related to service worker or not.

in any case, it seems to be the carbon ads json that is never loaded (shows as "(pending)" in the F12 tools' network tab). only once the browser gives up trying to load it is the JS on page load actually executed.

anyway, that's all tangential. the carousel swipe thing now seems to work nicely

**Comment by XhmikosR** [comment]
This no longer works for me with Firefox 63 devtools and say S9 emulation. It works fine with Chrome though.

**Comment by Johann-S** [comment]
@XhmikosR that's strange it works perfectly on my Android with Firefox

**Comment by XhmikosR** [comment]
Well a real touch device is not the same as the emulator :) That being said I don't know if we should even care about it. I just thought I'd report it since Firefox is my main browser and this worked in the previous patches.

**Comment by Johann-S** [comment]
I think we should test using BrowserStack here

**Comment by XhmikosR** [comment]
All right, let's merge this then. Great work everyone!

**Pull request merged**
